base/compiler/ssair/inlining2.jl

```diff
@@ -346,7 +346,62 @@ function ir_inline_item!(compact::IncrementalCompact, idx::Int, argexprs::Vector
     return_value
 end
 
-function ir_inline_unionsplit!(compact::IncrementalCompact, idx::Int,
+# Constraints are generally small, so a linear search is the bets option
+function find_constraint(val, constraints)
+    for i = 1:length(constraints)
+        if val === constraints[i][1]
+            return constraints[i][2]
+        end
+    end
+    return nothing
+end
+
+# Performs minimal backwards inference to catch a couple of interesting, common cases
+function minimal_backinf(compact, constraints, unconstrained_types, argexprs)
+    for i = 2:length(argexprs)
+        isa(argexprs[i], SSAValue) || continue
+        # Check if the argexpr is in the constraint list directly
+        c = find_constraint(argexprs[i], constraints)
+        if c !== nothing
+            unconstrained_types[i] = c
+        end
+        # For boolean values check for type predicates on any of the constraints
+        ut = unconstrained_types[i]
+        if ut === Bool
+            def = compact[argexprs[i]]
+            isa(def, Expr) || continue
+            if is_known_call(def, ===, compact)
+                v1, v2, = def.args[2:3]
+                c = find_constraint(v1, constraints)
+                if c !== nothing
+                    refined = egal_tfunc(c, compact_exprtype(compact, v2))
+                    if !(ut ⊑ refined)
```

Comment:
might be better to preserve the inference type, unless we've proven that we have a narrower bound (instead of checking if we're certain we don't have a wider bound):
```if refined ⊑ ut```